src/quorumnet/sn_network.cpp

```diff
@@ -0,0 +1,1019 @@
+}
+
+void SNNetwork::register_public_command(std::string command, std::function<void(SNNetwork::message &message, void *data)> callback) {
+    if (!commands_mutable)
+        throw std::logic_error("Failed to register public command: command must be added before constructing a SNNetwork instance");
+
+    commands.emplace(std::move(command), std::make_pair(std::move(callback), true));
+}
+
+std::atomic<int> next_id{1};
+
+/// We have one mutex here that is generally used once per thread: to create a thread-local command
+/// socket to talk to the proxy thread's control socket.  We need the proxy thread to also have a
+/// copy of it so that it can close them when it is exiting.
+std::mutex local_control_mutex;
+
+/// Accesses a thread-local command socket connected to the proxy's command socket used to issue
+/// commands in a thread-safe manner (without requiring a mutex).
+zmq::socket_t &SNNetwork::get_control_socket() {
+    // Maps the SNNetwork unique ID to a local thread command socket.
+    static thread_local std::map<int, std::shared_ptr<zmq::socket_t>> control_sockets;
+
+    auto it = control_sockets.find(object_id);
+    if (it != control_sockets.end())
+        return *it->second;
+
+    std::lock_guard<std::mutex> lock{local_control_mutex};
+    zmq::socket_t foo{context, zmq::socket_type::dealer};
+    auto control = std::make_shared<zmq::socket_t>(context, zmq::socket_type::dealer);
+    control->setsockopt<int>(ZMQ_LINGER, 0);
+    control->connect(SN_ADDR_COMMAND);
+    thread_control_sockets.push_back(control);
+    control_sockets.emplace(object_id, control);
+    return *control;
+}
+
+
+SNNetwork::SNNetwork(
+        std::string pubkey_, std::string privkey_,
+        const std::vector<std::string> &bind,
+        LookupFunc lookup,
+        AllowFunc allow,
+        WantLog want_log,
+        WriteLog log,
+        unsigned int max_workers)
+    : object_id{next_id++}, peer_lookup{std::move(lookup)}, allow_connection{std::move(allow)}, want_logs{want_log}, logger{log}, max_workers{max_workers}, pubkey{std::move(pubkey_)}, privkey{std::move(privkey_)}
+{
+    SN_LOG(trace, "Constructing SNNetwork, id=" << object_id << ", this=" << this);
+
+    if (bind.empty())
+        throw std::invalid_argument{"Cannot create a service node with no address(es) to bind"};
+
+    commands_mutable = false;
+
+    SN_LOG(info, "Initializing SNNetwork quorumnet listener with pubkey " << as_hex(pubkey));
+    assert(pubkey.size() == 32 && privkey.size() == 32);
+
+    if (max_workers == 0)
+        max_workers = 1;
+
+    {
+        // We bind here so that the `get_control_socket()` below is always connecting to a bound
+        // socket, but then proxy thread is responsible for the `command` socket.
+        command.bind(SN_ADDR_COMMAND);
+        proxy_thread = std::thread{&SNNetwork::proxy_loop, this, bind};
+
+        SN_LOG(warn, "Waiting for proxy thread to get ready...");
+        auto &control = get_control_socket();
```

Comment:
Seems like this can be done with a std::atomic instead of the back and forth manual synchronisation?

Reply:
Aside from avoiding a mutex (see above), ZMQ sockets are somewhat different from raw sockets in that zmq doesn't actually open an underlying connection until there is something to send on it--you don't have a guarantee just because you've opened it that the remote can reach you.  Ping-ponging through the socket is essentially giving us a guarantee that the socket is established and functioning (and is probably faster than the mutex required by an atomic, not that that particularly matters in this context).